test/jb_controller/mint_tokens_of.test.js

```diff
@@ -0,0 +1,441 @@
+      .withArgs(terminalSigner.address, projectOwner.address, PROJECT_ID, MINT_INDEX)
+      .returns(false);
+
+    await mockJbOperatorStore.mock.hasPermission
+      .withArgs(terminalSigner.address, projectOwner.address, 0, MINT_INDEX)
+      .returns(false);
+
+    await mockJbDirectory.mock.isTerminalDelegateOf
+      .withArgs(PROJECT_ID, terminalSigner.address)
+      .returns(true);
+
+    await mockJbFundingCycleStore.mock.currentOf.withArgs(PROJECT_ID).returns({
+      // mock JBFundingCycle obj
+      number: 1,
+      configuration: timestamp,
+      basedOn: timestamp,
+      start: timestamp,
+      duration: 0,
+      weight: 0,
+      discountRate: 0,
+      ballot: ethers.constants.AddressZero,
+      metadata: packFundingCycleMetadata({ pauseMint: 1, reservedRate: RESERVED_RATE }),
+    });
+
+    await expect(
+      jbController
+        .connect(terminalSigner)
+        .mintTokensOf(PROJECT_ID, AMOUNT_TO_MINT, beneficiary.address, MEMO, true, RESERVED_RATE),
+    )
+      .to.emit(jbController, 'MintTokens')
+      .withArgs(
+        beneficiary.address,
+        PROJECT_ID,
+        AMOUNT_TO_MINT,
+        MEMO,
+        RESERVED_RATE,
+        terminalSigner.address,
+      );
+
+    let newReservedTokenBalance = await jbController.reservedTokenBalanceOf(
+      PROJECT_ID,
+      RESERVED_RATE,
+    );
+    expect(newReservedTokenBalance).to.equal(AMOUNT_TO_MINT - AMOUNT_TO_RECEIVE);
+  });
+
+  it(`Should add the minted amount to the reserved tokens if reserved rate is 100%`, async function () {
+    const {
+      projectOwner,
+      beneficiary,
+      jbController,
+      mockJbFundingCycleStore,
+      mockTokenStore,
+      timestamp,
+    } = await setup();
+
+    await mockJbFundingCycleStore.mock.currentOf.withArgs(PROJECT_ID).returns({
+      // mock JBFundingCycle obj
+      number: 1,
+      configuration: timestamp,
+      basedOn: timestamp,
+      start: timestamp,
+      duration: 0,
+      weight: 0,
+      discountRate: 0,
+      ballot: ethers.constants.AddressZero,
+      metadata: packFundingCycleMetadata({ reservedRate: 10000 }),
+    });
+
+    await mockTokenStore.mock.totalSupplyOf.withArgs(PROJECT_ID).returns(0);
+
+    let previousReservedTokenBalance = await jbController.reservedTokenBalanceOf(
+      PROJECT_ID,
+      /*reservedRate=*/ 10000,
+    );
+
+    await expect(
+      jbController
+        .connect(projectOwner)
+        .mintTokensOf(PROJECT_ID, AMOUNT_TO_MINT, beneficiary.address, MEMO, true, 10000),
+    )
+      .to.emit(jbController, 'MintTokens')
+      .withArgs(beneficiary.address, PROJECT_ID, AMOUNT_TO_MINT, MEMO, 10000, projectOwner.address);
+
+    let newReservedTokenBalance = await jbController.reservedTokenBalanceOf(PROJECT_ID, 10000);
+
+    expect(newReservedTokenBalance).to.equal(previousReservedTokenBalance.add(AMOUNT_TO_MINT));
+  });
+
+  it(`Should substract the received amount to the reserved tokens if reserved rate is 0%`, async function () {
```

Comment:
hmmm, i think reserved tokens should not change if reserved rate == 0.

Reply:
Indeed @mejango :

https://github.com/jbx-protocol/juice-contracts-v2/blob/62fda9c81d6f820deede0a2a271f1cd6b8772fbb/contracts/JBController.sol#L817

With a reserved rate == 0, _unprocessed = total supply - processed and 
reservedToken = [ (unprocessed * 100) / (100 - 0) ] - unprocessed = 0

-> this is correct in implementation, I'm fixing the test, which should always returns 0 (never neg, don't know why I did this):
![image](https://user-images.githubusercontent.com/83670532/148677452-9ddad2e7-0351-4f58-a3af-952b24da931d.png)

Reply:
Fixed it like this : `expect(newReservedTokenBalance).to.equal(previousReservedTokenBalance).and.to.equal(0);`

Reply:
nice. was the test passing beforehand?

Reply:
the `to.equal(0)` bit depends on your setup script. If there is a previousReservedTokenBalance, this will remain.
i think we can get rid of the `and.to.equal(0)` for clarity perhaps?

Reply:
Yes, it was passing before (always equal to 0). Will remove the to.equal